README
Black Flower CAD v1.8.0 - 2011/06/19

OVERVIEW

  Description:
  ------------
  Black Flower CAD is an open source Computer Aided Dispatch / Computer 
  Aided Logging software package.  
  
  Contact Information:
  --------------------
  The current version of this software can be downloaded from:
    http://www.forlorn.net/cad/
  
  For bug reports, questions, etc, send email to:
    cad-info@forlorn.net
  

UPGRADE INSTRUCTIONS

   (For new installations, see INSTALLATION below.)

   Upgrading from 1.4.x or previous is not supported due to lack of need 
   and therefore nonexistant database upgrade scripts.  Contact us if this
   need exists in your environment.

   To upgrade from versions 1.5 or greater:

   * Compare cad.conf.example to your current cad.conf (or run initialize.sh
     if you do not yet have cad.conf.)  Create or adjust variable names as 
     necessary.
   * Run all database schema update scripts in data/updates subdirectory,
     sequentially by version number.



INSTALLATION
  
  Server Requirements:
  --------------------
  * MySQL database server v5.0 or greater
  * PHP version 5, with MySQL integration.
  * Apache 2.0+ webserver running in Prefork mode, with PHP integration.
  
  * Linux (Debian/CentOS) recommended for mission critical installations.
  * Microsoft Windows (XP/7) has also been used in development environments.
    -> IMPORTANT:  Production use on Microsoft Windows is NOT recommended,
       due to limitations of PHP running under Windows-based webservers and
       using Windows' Event Log emulation for syslog.
  
  
  Client Requirements:
  --------------------
  * 1024x768 (or greater) screen resolution required.
    -> 1280x1024 (or greater) resolution is recommended for best results.
  * Mozilla v1.5+ or Firefox v1.0+ web browser required.
    -> Internet Explorer IS EXPRESSLY NOT SUPPORTED.
    -> Safari and Chrome have been minimally tested, and cosmetic issues
       were noted.  They may or may not be fully functionally compatibile.
    -> Other browsers have not been tested and are not likely to work.

  * Client system time must be synchronized with CAD server system time.
  
  * Microsoft Windows (XP/7) is recommended for client operating system.
    -> Linux clients have also been tested successfully.
    -> Apple OS X has been minimally tested.
  
  
  Installation Quick Guide
  ------------------------
  1) Install the MySQL database server on the target server.
     *** Linux instructions coming later; refer to MySQL documentation ***
     *** Windows instructions coming later; refer to MySQL documentation ***

     When installing this server, assign and memorize a password to
     the MySQL "root" user account.  This will be required during CAD
     installation.

     Unless you know exactly what you are doing, it is STRONGLY advised 
     to allow MySQL (port 3306) connections *only* from localhost.
  
  2) Install Apache and PHP on the target server.
     *** Linux instructions coming later; refer to Apache/PHP docs ***
     *** Windows instructions coming later; refer to Apache/PHP docs ***

  3) Unpack the CAD distribution archive in the desired Web directory.

     Example (for Debian Linux -- HTML documents directory may be different
     on other operating systems):

     # Acquire adequate (root) permissions to write to destination, then:

       cd /var/www/
       tar zvxf ~/cad-<version>.tar.gz
       chown -R www-data.www-data cad-<version>
       ln -sf cad-<version> cad
  
  4) Configure the CAD system.

     If bash and perl are available on your Linux or UNIX server, change 
     to the CAD directory and execute the "initialize.sh" script.  

       cd /var/www/cad
       bash ./initialize.sh

     This will create and load the default SQL tables and otherwise 
     prepare the system for running.  You will be prompted to enter
     certain usernames and passwords.  Or, if you have a fair level of
     MySQL command line administration skills, read the installation script
     and initialize the system manually using the MySQL root account.

     The initialize.sh script does not currently error-detect for an 
     already existing database or user with specified names.  To run
     the script multiple times, you must first change or delete conflicting 
     data.  Be careful.
  
     If you are on a multiuser system, ensure that cad.conf is owned
     by a user/group that is readable only by the webserver, and change
     file permissions to match, so as to safeguard your passwords.

     *** Manual/Windows documentation to be added ***  
     
  5) Configure your Apache installation to serve the CAD application.

     The Apache configuration for your desired CAD HTML directory must
     permit the "AllowOverride Limit" configuration.  A ".htaccess" file
     is used to block access to sensitive application files.  If your
     Apache config does not permit this, update it and reload Apache
     at this time.
  
     Test this configuration by trying to load a file that should
     be forbidden by the .htaccess rules: 
       http://<server>/<path-to-CAD-application>/cad.conf

     This should NOT succeed.  If you ARE able to load this file,
     troubleshoot the Apache site config and provided .htaccess file.

  6) Synchronize the server and all clients to the same time.  
     See NTP discussion below.
  
  
  To use the CAD system:
  ----------------------
  On the server, start or restart Apache and MySQL as needed.
  On the client, load your web browser and go to the CAD URL, e.g.:

    http://<server>/<path-to-CAD-application>/

  (or https:// if you have configured your system for HTTPS.)
  
  Log in with the CAD administrator username and password created 
  during step 4 above.  The password is case sensitive.
  
  Click on the "Settings" tab, then "Edit Users" in the Administration 
  dialog, and create the required usernames and passwords for CAD users.

  An access level less than 10 should be used for normal users.  
  Access levels of 10 or greater create a system administrator account.  

  Day to day use of CAD should be done through normal, rather than
  system administrator, accounts.
  
  
  Best Common Practices
  ---------------------
  1)  NTP
  
  Time synchronization is required for coordination of timestamps 
  generated on the server with those generated on the clients.  
  This can be done by running a standalone (stratum 1) NTP server 
  if on a completely isolated network, or by connecting all systems 
  to the Internet NTP network.  In the Internet case it is 
  recommended to run a caching NTP server on the CAD server, and 
  synchronize all clients to that.  A popular freeware Windows NTP 
  client is available from http://www.oneguycoding.com/automachron/.
  
  2)  Syslog
  
  CAD will use the native system logging functionality on UNIX or 
  Windows systems.  On UNIX systems, CAD uses the local4 logging level,
  and obeys the severity filter as set in cad.conf.  On Windows, 
  *all* messages are logged to the Application section of the Event 
  Viewer, disregarding the severity filter as set in cad.conf.  
  
  3)  Network security
  
  Any network running CAD or other such systems should be kept 
  isolated from insecure network traffic such as the Internet.  
  The CAD server should be reachable from the CAD clients but they
  should not be reachable from untrusted (outside/Internet) systems.
  Any consumer-grade or better hardware firewall is sufficient for 
  purposes of CAD network security.  
  
  In an extremely small or field-deployable CAD installation, it is 
  possible to install the server on a system which will also be one
  of the clients of the installation.  If that is the only system 
  expected to use CAD, a software firewall on that host is sufficient 
  if configured correctly.  Software firewalls are also generally 
  recommended for any Microsoft Windows client or server systems 
  as best practice.  CAD is a cleanly client-server system, it does 
  not require network ingress access for connections _to_ clients.
  
  Ports required to be accessible on CAD server by the CAD clients
  for application communication:
  
    80 (http)   - Normal operation: HTTP on a secure network.
  *or*
   443 (https)  - If using secure HTTPS on an insecure network
  
  Other ports that may be used in a full network environment:
  
    53 (domain) - If using the CAD server as a DNS server.
    67 (bootps) - If using the CAD server as a DHCP server.
   123 (ntp)    - If using the CAD server as an NTP server.
  
  
  


CHANGES
CHANGES
-------

---------------------------
Version 1.12.3 - 2016-07-22
---------------------------
[bugfix] Fixed Log Viewer from not displaying any messages under certain
         conditions.
[bugfix] Improved the logic for compiling partial incident header 
         information into unit dispatch auto-pages.

----------------------------------------
Version 1.12.2 (unreleased) - 2015-09-01
----------------------------------------
[minor]  On staging screen, show the times when each unit was staged there.

---------------------------
Version 1.12.1 - 2015-08-31
---------------------------
[bugfix] Prevent multiple operators from releasing an attached unit 
         multiple times and by doing so toggling its status irrevocably 
         back 'Attached To Incident'.

---------------------------
Version 1.12.0 - 2015-08-31
---------------------------
[minor]  Added initial version of unit utilization reports.
[minor]  Upgraded to FPDF 1.7, with assorted changes to AddPage() and 
         Output() calls.
[bugfix] Cleaned up various issues with reports.
[minor]  Added unit staging functionality.

---------------------------
Version 1.11.0 - 2014-08-23
---------------------------

[bugfix] Fixed erroneous incident "Undispatched" status that could be 
         displayed if User A created an incident, User B took over the
         incident's read/write lock, but User A then attached a unit 
         prior to refreshing the window and losing their original read/write 
         lock to User B.  Resolved multiple related bug reports.
[minor]  Changed Incident window to display times as full date + time rather
         than just name of day + time, if a timestamp was from a different 
         calendar day.  Display is not very elegant currently, but is more 
         correct functionality.
[minor]  Implemented configurable permissions levels for administrative
         and supervisory functions.
[minor]  Implemented single-incident report.
[minor]  Restricted incident note entry from allowing browser form history.
[minor]  Tweaked incident list "Undispatched" flag display style for increased 
         visibility. (blink property doesn't work everywhere any more.)
[minor]  Fixed config-cleardb to work with Mysql 5.x constraints.

---------------------------
Version 1.10.1 - 2013-09-30
---------------------------

[minor]  Applied viewport setting to stop iPad client from continually 
         refreshing incidents pane bigger than one screen.
[minor]  Added unit response time reports.
[minor]  Enhanced unit reports by allowing range of start and end dates.
[minor]  Added unit daily summary reports.

---------------------------
Version 1.10.0 - 2013-07-18
---------------------------

[minor]  Allow filtering incident reports by incident type.
[minor]  Added ability to prevent closing (dispositioning) incidents unless 
         all assigned units are first manually released from it.  This is 
         an administrator-configurable feature:  in cad.conf, set:
           $FORCE_MANUAL_UNIT_RELEASE= 1;
[minor]  Added ability to require supervisors to review and approve the 
         details of a Dispositioned incident before it is fully Completed.
         This is an administrator-configurable feature:  in cad.conf, set:
           $SUPERVISOR_INCIDENT_REVIEW = 1;
[minor]  Now require cad.conf "$ALLOW_MANUAL_PAGE = 1;" to enter a page 
         to a unit manually from CAD.  Automatic page-out still works 
         regardless of this setting.
[bugfix] Disable channel assignment functionality on a closed incident.
[bugfix] Disable note "submit" button on a closed incident.


--------------------------
Version 1.9.2 - 2013-04-17
--------------------------

[minor]  Importing minor additions for icon definitions and preloaded 
         incident type 'COURTESY TRANSPORT', changes were made but not 
         committed in August.

--------------------------
Version 1.9.1 - 2012-08-25
--------------------------

[minor]  Added ability to resolve a call as a duplicate of another call.

[minor]  Don't show duplicate/training calls in status totals.

[minor]  Focus cursor to unit name field on new unit screen.

[minor]  Improved cosmetics of incidents reports and unit reports.

--------------------------
Version 1.9.0 - 2012-08-21
--------------------------
[major]  THE PASSWORD HASH FORMAT HAS CHANGED AS OF 1.9.0.  All passwords 
         generated on previous versions must be reset per this example
         (starting with the Administrator account for convenience):

  1. Set $DEBUG=1 in cad.conf, 
  2. Attempt to log in as Administrator (which will fail), 
  3. Observe the hash displayed on the login failure screen,
  4. Update the MySQL users table with the displayed hash: 
     mysql>  update cad.users set password='<hash>' where id=1;
  5. Log in as Administrator with your new password.
  6. Go to the System Admin tab and change all other users' passwords.
  7. Set $DEBUG=0 in cad.conf.

[major]  Implemented channel assignments to incidents.  Channels may
         be configured by the sysadmin to be available or unavailable
         (for reference only), repeater or simplex (for display purposes),
         then assigned by a dispatcher in the edit_incident window.

[major]  Implemented optional read-only mode and associated config settings.
         If read-only settings are not defined, the app will remain wide open
         (in read-write mode) for all connections.
         
[minor]  Removed green interstitial process screen from new incident creation.

[minor]  Users are now locked out after repeated incorrect login attempts
         (default limit = 5 attempts, configurable in cad.conf.)

[minor]  Implemented forced password expiry.

[bugfix] Fixed bugs in error checking for System Admin > Edit Users.


--------------------------
Version 1.8.2 - 2011-09-03

unit.php
<?php
  $subsys="unit";
  
  require_once('db-open.php');
  include('local-dls.php');
  require_once('session.inc');

  if (isset($_POST['unit'])) {
      $unit = strtoupper(MysqlClean($_POST,'unit',20));
      if (isset($_POST['status'])) $status = MysqlClean($_POST,'status',30); else $status = "";
      if (isset($_POST['status_comment'])) $status_comment = MysqlClean($_POST,'status_comment',255); else $status_comment= "";
      if (isset($_POST['type'])) $type = MysqlClean($_POST,'type',20); else $type= "Unit";
      if (isset($_POST['role'])) $role = MysqlClean($_POST,'role',20); else $role= "Other";
      if (isset($_POST['personnel'])) $personnel = MysqlClean($_POST,'personnel',100); else $personnel= "";
  }

  if (isset($_POST['saveunit'])) {
    if (isset($_POST['new-unit-entered'])) {
      $unit = strtoupper(MysqlClean($_POST,'unit',20));
      $pattern = "/[\\/[\]'!@#$\^%&*()+=,;:{}|<>~`?\"]/";
      $replacement = "";
      if (preg_match($pattern, $unit)) {
        die('Bad characters in name: '.$unit. "\n  Only letters, numbers, space, dash or underscore are valid characters.\n  Use your browser Back feature to resolve the problem and try again.");
      }
      // update status
      if ($personnel != "") 
        $query = "INSERT INTO messages (ts, unit, message) VALUES (NOW(), '$unit', 'Unit created - personnel: $personnel')";
      else
        $query = "INSERT INTO messages (ts, unit, message) VALUES (NOW(), '$unit', 'Unit created.')";
      mysql_query($query) or die("In query: $query\nError: " . mysql_error());

      // update units
      // TODO: sanity check $unit input characters here?
      $query = "INSERT INTO units (unit, status, status_comment, type, role, personnel, update_ts) VALUES ('$unit', '$status', '$status_comment', '$type', '$role', '$personnel', NOW())";
      mysql_query($query) or die("In query: $query\nError: " . mysql_error());
    }

    else {
      // update status and personnel notes 
      if ($_POST['status'] <> $_POST['previous_status']) {
        $query = "INSERT INTO messages (ts, unit, message) VALUES (NOW(), '$unit', 'Status change: $status";
        if ($_POST['previous_status'] <> "") {
              $query .= " (was: ".MysqlClean($_POST, 'previous_status', 200).")";
        }
        $query .= "')";
        mysql_query($query) or die("In query: $query\nError: " . mysql_error());
  
        if ($_POST['previous_status'] == 'Attached to Incident') {
          $query = "UPDATE incident_units SET cleared_time=NOW() WHERE unit='$unit' AND cleared_time IS NULL";
          mysql_query($query) or die("In query: $query\nError: " . mysql_error());
        }
      }
      if ($_POST['personnel'] <> $_POST['previous_personnel']) {
        $query = "INSERT INTO messages (ts, unit, message) VALUES (NOW(), '$unit', 'Personnel change logged: $personnel')";
        mysql_query($query) or die("In query: $query\nError: " . mysql_error());
      }

      // update units
      if ($_POST['status'] <> $_POST['previous_status'])
         $fragment = "status='$status', update_ts=NOW(),";
      else $fragment="";
      $query = "UPDATE units SET $fragment status_comment='$status_comment', type='$type', role='$role', personnel='$personnel' WHERE unit='$unit'";
      mysql_query($query) or die("In query: $query\nError: " . mysql_error());
    }
    print "<SCRIPT LANGUAGE=\"JavaScript\"> window.opener.location.reload(); self.close()</SCRIPT>";
  }

  elseif (isset($_POST["deleteunit"])) {
    if (isset($_POST['deleteforsure'])) {
      $query = "DELETE FROM units WHERE unit='".MysqlClean($_POST,"unit",20)."'";
      mysql_query($query) or die("In query: $query<br>\nError: " . mysql_error());
      $query = "INSERT INTO messages (ts, unit, message) VALUES (NOW(), '$unit', 'Unit deleted.')";
      mysql_query($query) or die("In query: $query<br>\nError: " . mysql_error());
      print "<SCRIPT LANGUAGE=\"JavaScript\"> window.opener.location.reload(); self.close()</SCRIPT>";
    }
    else {
      $_GET['unit'] = MysqlClean($_POST,'unit',20);
      $unit = $_GET['unit'];
      $newunit = 0;
      $unitquery = "SELECT * from units where unit = '$unit'";
      $unitresult = mysql_query($unitquery) or die("unit query failed:" . mysql_error());
      $unitline = mysql_fetch_array($unitresult, MYSQL_ASSOC) or die ("unit not found in table");
	    mysql_free_result($unitresult);
    }
  }
  elseif (isset($_GET["new-unit"])) {
    $unitline["unit"] = "";
    $unit = "";
    $newunit = 1;
    $unitline["status"] = "(new unit)";
    $unitline["role"] = "Other";
    $unitline["type"] = "Unit";
    $unitline["status_comment"] = "";
    $unitline["personnel"] = "";
    $unitline["update_ts"] = "";
  }
  elseif (isset($_GET["unit"])) {
    $unit = MysqlClean($_GET,"unit",20);
    $newunit = 0;
    $unitquery = "SELECT * from units where unit = '$unit'";
    $unitresult = mysql_query($unitquery) or die("unit query failed:" . mysql_error());
    $unitline = mysql_fetch_array($unitresult, MYSQL_ASSOC) or die ("unit not found in table");
	  mysql_free_result($unitresult);
  }
  else
    die ('Unknown options to unit.php page load.');

/*--------------------------------------------------------------------------------------------*/?> 

<HTML>
<HEAD>
  <TITLE>Dispatch :: Unit Details</TITLE>
  <LINK REL=StyleSheet HREF="style.css" TYPE="text/css" MEDIA=screen>
  
</HEAD>
<BODY vlink=blue link=blue alink=cyan>

<FONT face="tahoma,ariel,sans">
<FORM name="myform" action="unit.php" method="post">

<?php if (!$newunit) print " <b>Unit: $unit </b>"; else print " <b>Creating a New Unit</b>";?>

  <p>
  <table width=660>
  <tr>
  <td width=20>&nbsp;</td>
  <td colspan=2 bgcolor="#aaaaaa" width=580>
     <table cellpadding=2 cellspacing=0 width=100%> 
     <tr>

  <?php 
  if ($newunit) {
    print "<td class=\"message\"><b>Unit name</b></td>\n";
    print "<td colspan=\"3\" class=\"message\"><input type=\"text\" name=\"unit\"><input type=\"hidden\" name=\"new-unit-entered\"></td>\n";
    print "</tr>\n<tr>\n";
    print "</tr>\n<tr>\n";
  }
  else {
    print "<input type=\"hidden\" name=\"unit\" value=\"".$unit."\">";
  }
  ?>

       <td class="message" width=150 STYLE="width: 150px">Status</td> 
       <td class="message"> 
         <select name="status" accesskey="s">
    <?php /*--------------------------------------------------------------------------------------*/

       $statusset=0;
       $statusquery = "SELECT * from status_options";
       $statusresult = mysql_query($statusquery) or die ("status query failed:" . mysql_error());
       while ($line = mysql_fetch_array($statusresult, MYSQL_ASSOC)) {
         echo "        <option ";
         if (!strcmp($line["status"], $unitline["status"])) {
	         $statusset=1;
	         echo "selected ";
	       }
	       echo "value=\"". $line["status"]."\">". $line["status"]."</option>\n";
       }
       if (!$statusset) {
         echo "        <option selected value=\"none\">\n";
       }
       print "</select>\n";
	     mysql_free_result($statusresult);
    /*--------------------------------------------------------------------------------------------*/ ?>

<input type=hidden name="previous_status" value="<?php print $unitline["status"] ?>">
</td>

<td class="message" width=50>Type</td>
 <td class="message" > <select width="100" STYLE="width: 100px" name="type" accesskey="t">

     <?php /*--------------------------------------------------------------------------------------*/
       $avail_types = array('Unit', 'Individual', 'Generic');
       if (array_search($unitline["type"], $avail_types) === FALSE) {
		       print "<option selected value=\"\"></option>\n";
       }
       foreach ($avail_types as $type) {
         print "<option ";
         if ($unitline["type"] == $type) print "selected ";
         print "value=\"$type\">$type</option>\n";
       }
    /*--------------------------------------------------------------------------------------------*/ ?>

 </select>
</tr>
<tr>
 <td class="message" align=right>Updated:</td> 
 <td class="message" align=right width=150><?php print dls_utime($unitline["update_ts"])?> </td>
             
 <td class="message" align=right>Branch</td>
 <td class="message" align=right> <select width="100" STYLE="width: 100px" name="role" accesskey="r">

       <?php /*--------------------------------------------------------------------------------------*/
         $avail_roles = array('Fire', 'Medical', 'Comm', 'MHB', 'Admin', 'Other');
         if (array_search($unitline["role"], $avail_roles) === FALSE) {
		         print "<option selected value=\"\"></option>\n";
         }
         foreach ($avail_roles as $role) {
           print "<option ";
           if ($unitline["role"] == $role) print "selected ";
           print "value=\"$role\">$role</option>\n";
         }
		  /*--------------------------------------------------------------------------------------------*/ ?>
		</select>
		</td></tr>

   <tr>
     <td class="message"  width=90>Comment</td> 
     <td class="message" colspan="5"> <input type="text" accesskey="c" maxlength="250" size="80" name="status_comment" value="<?php print $unitline["status_comment"]?>"> </td>
   </tr>

   <tr>
     <td class="message"  width=90>Personnel</td> 
     <td class="message" colspan="5"> 
         <input type="text" accesskey="c" maxlength="250" size="80" name="personnel" value="<?php print $unitline["personnel"]?>"> 
         <input type="hidden" name="previous_personnel" value="<?php print $unitline["personnel"]?>">
     </td>
   </tr>
   <?php if ($unitline["type"] == "Generic") print "<tr>\n<td class=\"message\" colspan=\"6\"><b>Note: As a generic unit, multiple instances of this unit may be simultaneously assigned to separate incidents.</b></td></tr>" ?>

   
   </table>
   </td>
 </tr>
 <tr><td></td></tr>
 <tr><td></td>
     <td><input type="submit" name="saveunit" value="Save"> <input type="reset" value="Cancel" onClick="window.opener.location.reload(); self.close()"></td>
     <?php 
     if ($newunit) {
       print "</tr>\n";
    }
    else {
       print "<td align=right><input type=\"submit\" name=\"deleteunit\" value=\"Delete Unit\"></td>\n";
       print "</tr>\n";
       if (isset($_POST["deleteunit"])) {
          print "<tr><td colspan=2><td align=right><font size=\"-1\"><blink>Confirm</blink> delete unit? &nbsp; </font> <input type=\"checkbox\" name=\"deleteforsure\"></td></tr>";
       }
       else {
          print "<tr><td colspan=2><td align=right><font color=\"lightgray\" size=\"-1\"><i>Confirm delete unit? &nbsp; </i></font><input type=\"checkbox\" disabled name=\"deleteforsure\"></td></tr>"; 
       }
    }
  ?>
  </table>
  </form>

<?php if (!$newunit) { ?>
<b>Last 10 Messages</b><br>
  <table><tr><td width=20></td><td bgcolor="#aaaaaa">
  <table cellpadding=2 cellspacing=1> <tr>
    <td class="message">Time</td>
    <td class="message">Message</td>
  </tr>

  <?php

     $rowquery = "SELECT * FROM messages WHERE unit = '$unit' AND deleted=0 ORDER BY oid DESC LIMIT 10";
     $rowresult = mysql_query($rowquery) or die("row Query failed : " . mysql_error());

     while ($line = mysql_fetch_array($rowresult, MYSQL_ASSOC)) {
        echo "\t<tr>\n";
        $td = "<td class=\"message\">"; 

        echo $td, dls_utime($line["ts"]), "</td>";
        echo $td, $line["message"], "</td>";
        echo "\t</tr>\n";
     }
     mysql_free_result($rowresult);
     mysql_close($link);
   ?>

  </table>
  </table>
  <?php } ?>

</body>
</html>


